domain/cloudimagemetadata/state/state.go

```diff
@@ -0,0 +1,344 @@
+				Arch:            m.ArchitectureName,
+				VirtType:        m.VirtType,
+				RootStorageType: m.RootStorageType,
+				RootStorageSize: m.RootStorageSize,
+				Source:          m.Source,
+			},
+			Priority:     m.Priority,
+			ImageId:      m.ImageID,
+			CreationTime: m.CreatedAt,
+		}
+	}
+	return result, nil
+}
+
+// AllCloudImageMetadata retrieves all cloud image metadata from the database without applying any filter criteria.
+// Returns a slice of [cloudimagemetadata.Metadata] and any error encountered during the retrieval process.
+func (s *State) AllCloudImageMetadata(ctx context.Context) ([]cloudimagemetadata.Metadata, error) {
+	result, err := s.FindMetadata(ctx, cloudimagemetadata.MetadataFilter{})
+	if errors.Is(err, cloudmetadataerrors.NotFound) {
+		return nil, nil
+	}
+	return result, errors.Trace(err)
+}
+
+// tryCleanUpExpiredMetadata removes metadata records from the cloud_image_metadata table that are older than [ExpirationDelay].
+//
+// Custom images doesn't expire.
+func (s *State) tryCleanUpExpiredMetadata(ctx context.Context) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	deleteQuery := fmt.Sprintf(`
+DELETE FROM cloud_image_metadata 
+WHERE source != '%s' AND  
+created_at < '%s';`, CustomSource, time.Now().Add(-ExpirationDelay).Format(time.DateTime))
+	deleteMetadataStmt, err := sqlair.Prepare(deleteQuery)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, deleteMetadataStmt).Run()
+		return err
+	}))
+}
+
+// ensuresArchitecturesLoaded initializes architecture data by loading it from the database if not already done.
+// architecture are readonly, inserted at the creation of the DB, so it should be ok to load them once at the start
+// of the controller.
+func (s *State) ensuresArchitecturesLoaded(ctx context.Context) (err error) {
+	s.archesOnce.Do(func() {
+		defer func() {
+			if err != nil {
+				s.archesOnce = sync.Once{} // reset Once since something has been wrong during the init process
+				s.architectureIDsByName = nil
+			}
+		}()
+
+		var db database.TxnRunner
+		db, err = s.DB()
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		}
+
+		var loadArchsStmt *sqlair.Statement
+		loadArchsStmt, err = sqlair.Prepare(`SELECT &architecture.* FROM architecture;`, architecture{})
+
+		var archs []architecture
+		err = errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+			return tx.Query(ctx, loadArchsStmt).GetAll(&archs)
+		}))
+		if err != nil {
+			return
+		}
+
+		s.architectureIDsByName = make(map[string]int, len(archs))
+		for _, arch := range archs {
+			s.architectureIDsByName[arch.Name] = arch.ID
+		}
+	})
+
+	return err
+}
+
+// validateMetadata checks the required fields in the provided Metadata and returns an error if any field is missing.
+func validateMetadata(m cloudimagemetadata.Metadata) error {
```

Comment:
Move to the service layer.

Reply:
Right, I will remove it for now and reintroduce it in the next PR.